src/Haskell/Ide/Engine/Plugin/Base.hs

```diff
@@ -104,26 +105,31 @@ hieGhcDisplayVersion :: String
 hieGhcDisplayVersion = compilerName ++ "-" ++ VERSION_ghc
 
 getProjectGhcVersion :: IO String
-getProjectGhcVersion = do
-  isStackProject   <- doesFileExist "stack.yaml"
-  isStackInstalled <- isJust <$> findExecutable "stack"
-  if isStackProject && isStackInstalled
-    then do
-      L.infoM "hie" "Using stack GHC version"
-      catch (tryCommand "stack ghc -- --numeric-version") $ \e -> do
-        L.errorM "hie" $ show (e :: SomeException)
-        L.infoM "hie" "Couldn't find stack version, falling back to plain GHC"
-        getGhcVersion
-    else do
-      L.infoM "hie" "Using plain GHC version"
-      getGhcVersion
+getProjectGhcVersion =
+  onException
+    (asum [tryStackProject, tryCabalProject, trySystemGhc])
+    (error "Couldn't find GHC version through stack, cabal, nor system GHC.")
 
   where
-    getGhcVersion = do
+    tryStackProject = do
+      isStackProject   <- doesFileExist "stack.yaml"
+      isStackInstalled <- isJust <$> findExecutable "stack"
+      guard $ isStackProject && isStackInstalled
+      L.infoM "hie" "Trying stack GHC version"
+      tryCommand "stack ghc -- --numeric-version"
+
+    tryCabalProject = do
+      isCabalProject <- doesFileExist "cabal.project"
+      isCabalInstalled <- isJust <$> findExecutable "cabal"
+      guard $ isCabalProject && isCabalInstalled
+      L.infoM "hie" "Trying cabal GHC version"
+      tryCommand "cabal v2-exec ghc -- --numeric-version"
```

Comment:
For older installed `cabal-install` versions `v2-exec` won't be understood.

I wonder if we should make a `runQuietCabal` which does `cabal --version`, and depending on the result uses the `vX-` prefixes or not.

Reply:
We can use `new-*` for now. Because applying `v2-` instead of `new-` should be done wider (wherever we do `build`, `configure`, etc.